app/util/links.ts

```diff
@@ -71,3 +73,117 @@ export function getStatusLink(urlRoot: string, jobID: string, rel: string): Link
     type: 'application/json',
   };
 }
+
+/**
+ * Given a job event, return the JobLink representing the corresponding event.
+ * @param event - the event being triggered by the JobLink
+ * @param jobID - the jobID that this link is for
+ * @param urlRoot - the root url to use in the link
+ * @param isAdmin - whether to include /admin in the link
+ * @returns - the job link for the job event
+ */
+function getLinkForJobEvent(
+  event: JobEvent, 
+  jobID: string, 
+  urlRoot: string, 
+  isAdmin = false,
+): JobLink {
+  const adminPath = isAdmin ? '/admin' : '';
+  switch (event) {
+    case JobEvent.RESUME:
+      return new JobLink({
+        title: 'Resumes the job.',
+        href: `${urlRoot + adminPath}/jobs/${jobID}/resume`,
+        type: 'application/json',
+        rel: 'resumer',
+      });
+    case JobEvent.SKIP_PREVIEW:
+      return new JobLink({
+        title: 'Skips preview and runs the job.',
+        href: `${urlRoot + adminPath}/jobs/${jobID}/skip-preview`,
+        type: 'application/json',
+        rel: 'preview-skipper',
+      });
+    case JobEvent.CANCEL:
+      return new JobLink({
+        title: 'Cancels the job.',
+        href: `${urlRoot + adminPath}/jobs/${jobID}/cancel`,
+        type: 'application/json',
+        rel: 'canceler',
+      });
+    case JobEvent.PAUSE:
+      return new JobLink({
+        title: 'Pauses the job.',
+        href: `${urlRoot + adminPath}/jobs/${jobID}/pause`,
+        type: 'application/json',
+        rel: 'pauser',
+      });
+  }
+}
+
+/**
+ * Return a set of JobEvents representing the actions that are currently available
+ * to a user for a particular job.
+ * Note that this only returns actions that users can precipitate via state change links
+ * (e.g. JobEvent.FAIL will not be returned).
+ * @param job - the job to return valid actions (JobEvents) for
+ * @returns a set of JobEvent
+ */
+export function getLinkRelevantJobEvents(job: Job): Set<JobEvent> {
+  const transitions: [JobEvent, JobStatus][] = [
+    // [event, resultant status]
+    [JobEvent.CANCEL, JobStatus.CANCELED],
+    [JobEvent.PAUSE, JobStatus.PAUSED], 
+    [JobEvent.RESUME, JobStatus.RUNNING], 
+  ];
+  if (job.status === JobStatus.PREVIEWING) {
+    // This may be a valid transition for other states, but we
+    // are only interested in it when the job is currently previewing.
+    // e.g. skipping preview is valid for a paused job but doesn't make sense to a user
+    transitions.push([ JobEvent.SKIP_PREVIEW, JobStatus.RUNNING ]);
+  }
+  const validEvents = new Set<JobEvent>();
+  for (const [event, newStatus] of transitions) {
+    try {
+      validateTransition(job.status, newStatus, event);
+      validEvents.add(event);
+    } catch {}
```

Comment:
Empty catch blocks are usually a bad sign (I know, I'm guilty too) as they indicate that you are using try/catch for normal control flow not just exceptional things. One good solution might be to refactor the `vaidateTransition` function into two functions, one that returns an error message if there is an error (or null if not) and one that wraps the other functions and throws an exception if the error message is not null. Then you could just call the first function here and check it's return value.

Reply:
Thanks for the explanation, fixed!